### ISSUE TITLE
[TF 2.0] Feature request: let autograph accept collection inputs

### ISSUE THREAD

**Issue by David-Mao** [tensorflow/tensorflow#32457]
**System information**
- TensorFlow version (you are using): 2.0.0-dev20190901
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**

Currently it seems to me that there is no way to convert a function to graph if the input of the function is a variant length list of elements (except the simplest case when the elements themselves are all tensors with the same shape, for which we can use a TensorArray). 

Assume we have a function that runs well in graph mode for a certain type of input, say a tensor tuple containing 2 tensors with arbitrary shape:

```
@tf.function(input_signature=[(tf.TensorSpec(None), tf.TensorSpec(None))])
def do_something(input):
    ....
    return some_result
```

We want to have another function that takes undetermined number of `input`s and merge the result, so we have:

```
def merge(list_of_inputs):
    return tf.reduce_sum(tf.stack([do_something(input) for input in list_of_inputs]))
```

This function works well in eager mode, but we cannot add a `@tf.function` decorator to it, because `list_of_inputs` has no fixed length and the function will retrace for every calls, which is very slow. Even if we can pad the `list_of_inputs` to have the same length, it's still not working as the shapes of the elements might vary across different runs, so retracing will still be triggered.

So it would be great if we have something like `tf.Array`and `tf.ArraySpec`, which is defined as an array of objects that share the same specs. Then we can give the `merge` function a decorator like:

```
@tf.function(input_signature=[
    tf.ArraySpec((tf.TensorSpec(None), tf.TensorSpec(None)))])
def merge(list_of_inputs):
    return tf.reduce_sum(tf.stack([do_something(input) for input in list_of_inputs]))
```
 and run it without retracing.

**Will this change the current api? How?**

As described above, we propose to add `tf.Array`and `tf.ArraySpec`.

**Who will benefit with this feature?**

Anyone uses autograph will benefit. It would be much easier to write code that handles the case when a data example contains a set/list of variant number of objects, which is very common in NLP / vision / knowledge graph and so on.

**Comment by mdanatg** [comment]
To make sure I understand, you are suggesting adding support for heterogeneous dynamic lists, kind of like a `TensorArray` with elements of unknown shape? There are indeed primitive ops that should allow building such a type, although the gradients would probably become a bit more complicated.

Could you elaborate a bit on the ArraySpec piece a bit - you defined it as an array of objects that share the same specs - in what way are the specs shared? In the example you gave, does merge take an arbitrary-length list of tuples, but the tuples all have the same shapes?

**Comment by David-Mao** [comment]
> Could you elaborate a bit on the ArraySpec piece a bit - you defined it as an array of objects that share the same specs - in what way are the specs shared? In the example you gave, does merge take an arbitrary-length list of tuples, but the tuples all have the same shapes?

Yes, `merge` should take an arbitrary-length list of tuples, and the tuples all have the same signatures. I mention signatures rather than shapes here because signatures allow unknown dimension (i.e. the `None` value in `tf.shape()`), which is more flexible.

**Comment by mdanatg** [comment]
Thanks! This makes sense. Essentially, this would be a more generic version of TensorArray that can hold structures, not just single Tensors.

Right now, there are a few ways work with this kind of data (structures of TensorArray, Datasets), but neither as as practical as an actual Array class. So I think this is a good feature request.

**Comment by mmjb** [comment]
Just want to chime in here: This feature request is crucial to getting decent performance out of TF2 for Graph Neural Network implementations.

In the GNN setting, such a list of inputs are shape [None, 2] tensors of adjacency lists for different edge types. The number of edge types are fixed for a given instance of the GNN model, but of course not an inherent property of the model class. See https://github.com/microsoft/tf-gnn-samples/blob/master/gnns/rgcn.py for an example of a TF1 implementation of this, illustrating how the computation graph would look.

The performance difference between that example to a TF2 implementation without AutoGraph is roughly ~5x (!), which frankly makes TF2 unusable for serious uses of such models.

The solution proposed by @David-Mao would work well for this case: Practically, this would just mean that the number of elements in the input list (or tuple) would become part of the shape of inputs; and hence `foo(x, [v_0, ..., v_n], y)` should be handled no different than `foo(x, v_0, ..., v_n, y)` for implementation purposes, but with the advantage that the model author doesn't need to change `foo` every time `n` changes. Changing `n` should in most cases cause re-tracing (though interestingly enough, @David-Mao's example above wouldn't require that).

**Comment by mmjb** [comment]
Alright, we found "fun" workaround for this (left in the issue for anyone else encountering this problem).

Existing code:
```
class GNNInput(NamedTuple):
    """Input named tuple for the GNN."""

    node_features: tf.Tensor
    adjacency_lists: Tuple[tf.Tensor, ...]

class GNN(tf.keras.layers.Layer):
    [...]
    def build(self, tensor_shapes: GNNInput):
        [...]
        super().build(tensor_shapes)

    def call(self, inputs: GNNInput, training: bool = False):
        [...]
```

New `build`, enabling autograph with build-time constant shapes:
```
        call_input_spec = (
            GNNInput(
                node_features=tf.TensorSpec(
                    shape=tf.TensorShape((None, initial_node_features_shape[1])), dtype=tf.float32
                ),
                adjacency_lists=tuple(
                    tf.TensorSpec(shape=tf.TensorShape((None, 2)), dtype=tf.int32)
                    for _ in range(len(adjacency_list_shapes))
                ),
            ),
            tf.TensorSpec(shape=(), dtype=tf.bool),
        )
        setattr(self, "call", tf.function(func=self.call, input_signature=call_input_spec))
```

This is clearly not pretty, but makes things go reasonably fast. Clearly, all this malarkey shouldn't be necessary, and a `@tf.function` should just magically work, but there you go.

**Comment by mdanatg** [comment]
@mmjb thanks for sharing, it's a clever workaround!

As a side note, a few related APIs which generalize this notion to "any structure of tensors of corelated shapes" is [StructuredTensor](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md), (work-in-progress) and [RaggedTensor](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/ragged) (available, but not fully integrated with `tf.function` yet).

@kkimdev FYI

**Comment by mmjb** [comment]
Thanks for pointing to the StructuredTensor RFC; that will clearly help for such models. We've been trying to use `NamedTuple`s for related things (see the sample above), but something that's better-integrated with the rest of tensorflow would be much appreciated...

**Comment by David-Mao** [comment]
@mmjb To make sure I understand correctly, in your workaround `len(adjacency_list_shapes)` is fixed at the time of model building and cannot be changed for each different calls of `call()`, right? You didn't explicitly define `adjacency_list_shapes` in the code snippet so I don't know how that's passed...

**Comment by mmjb** [comment]
@David-Mao Yes, the list length is fixed at model-building time and is the same for all `call`s.

**Comment by David-Mao** [comment]
@mmjb Thanks.
In my desired use case I would like to see the list length varies with calls, so this workaround doesn't work for me :(

**Comment by kumariko** [comment]
@David-Mao We could see you are using TF v2.0.0 which is not actively supported ,please refer this [comment](https://github.com/tensorflow/tensorflow/issues/52915#issuecomment-961288304). Could you please try to upgrade TF version to latest TF v2.7.0 and let us know ? Thanks!
